File: cpp/inter/expr_tests.cpp
#include <gtest/gtest.h>

#include "cpp/inter/access.hpp"
#include "cpp/inter/and.hpp"
#include "cpp/inter/arithmetic-operator.hpp"
#include "cpp/inter/expression.hpp"
#include "cpp/inter/identifier.hpp"
#include "cpp/inter/node.hpp"
#include "cpp/inter/not.hpp"
#include "cpp/inter/or.hpp"
#include "cpp/inter/relational.hpp"
#include "cpp/inter/temporary.hpp"
#include "cpp/inter/unary.hpp"

#include "cpp/lexer/token.hpp"
#include "cpp/lexer/word.hpp"
#include "cpp/lexer/type.hpp"

#include <string>
#include <sstream>
#include <vector>

using inter::Access;
using inter::And;
using inter::Arithmetic;
using inter::Expression;
using inter::Identifier;
using inter::Node;
using inter::Not;
using inter::Or;
using inter::Relational;
using inter::Temporary;
using inter::UnaryOperator;

using lexer::Word;
using lexer::Token;

using symbols::Type;

namespace {
struct TestCase {
  std::shared_ptr<Expression> expr;
  std::string str;
  std::string gen;
  std::string red;
};
}

TEST(TestExpr, InterTests) {
  std::vector<TestCase> test_cases{
    TestCase{
      Identifier::create(Word::create("example", Token::kIdentifier), Type::integer, 4),
      "example",
      "",
      "",
    },
    TestCase{
      Temporary::create(Type::integer),
      "t1",
      "",
      "",
    },
    TestCase{
      Arithmetic::create(
        Token::create('+'),
        Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
        Identifier::create(Word::create("y", Token::kIdentifier), Type::integer, 4)
      ),
      "x + y",
      "",
      "\tt1 = x + y\n",
    },
    TestCase{
      UnaryOperator::create(
        Token::create('-'),
        Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4)
      ),
      "- x",
      "",
      "\tt1 = - x\n",
    },
    TestCase{
      Access::create(
        Identifier::create(Word::create("arr", Token::kIdentifier), Type::real, 4),
        Identifier::create(Word::create("x", Token::kIdentifier), Type::integer, 4),
        Type::real
      ),
      "arr[ x ]",
      "",
      "\tt1 = arr[ x ]\n",
    },
    TestCase{
      Or::create(
        Word::or_word,
        Identifier::create(Word::create("x", Token::kIdentifier), Type::boolean, 4),
        Identifier::create(Word::create("y", Token::kIdentifier), Type::boolean, 4)
      ),
      "x || y",
      "\tif x goto L3\n\tiffalse y goto L1\nL3:\tt1 = true\n\tgoto L2\nL1:\tt1 = false\nL2:",
      "",
    },
    TestCase{
      And::create(
        Word::and_word,
        Identifier::create(Word::create("x", Token::kIdentifier), Type::boolean, 4),
        Identifier::create(Word::create("y", Token::kIdentifier), Type::boolean, 4)
      ),
      "x && y",
      "\tiffalse x goto L1\n\tiffalse y goto L1\n\tt1 = true\n\tgoto L2\nL1:\tt1 = false\nL2:",
      "",
    },
    TestCase{
      Relational::create(
        Word::equal,
        Identifier::create(Word::create("x", Token::kIdentifier), Type::boolean, 4),
        Identifier::create(Word::create("y", Token::kIdentifier), Type::boolean, 4)
      ),
      "x == y",
      "\tiffalse x == y goto L1\n\tt1 = true\n\tgoto L2\nL1:\tt1 = false\nL2:",
      "",
    },
  };

  for (const auto &tc : test_cases) {
    Node::reset_labels();
    Temporary::reset_temp_count();

    ASSERT_EQ(tc.expr->to_string(), tc.str);
    std::stringstream ss;
    tc.expr->gen(ss);
    ASSERT_EQ(ss.str(), tc.gen);
    ss.str("");
    tc.expr->reduce(ss);
    ASSERT_EQ(ss.str(), tc.red);
  }
}

File: java/com/dragon/ast/ExprTests.java
package com.dragon.ast;

import static org.junit.Assert.assertEquals;

import java.util.Arrays;

import com.dragon.lexer.Tag;
import com.dragon.lexer.Token;
import com.dragon.lexer.Word;
import com.dragon.symbols.Type;

import org.junit.Test;

public class ExprTests {
  private class TestCase {
    public Expr expr;
    public String str;
    public String gen;
    public String red;

    public TestCase(Expr expr, String str, String gen, String red) {
      this.expr = expr;
      this.str = str;
      this.gen = gen;
      this.red = red;
    }
  }

  @Test
  public void exprTests() {
    TestCase[] testCases = {
      new TestCase(
        new Id(new Word("example", Tag.ID), Type.Int, 4),
        "example",
        "",
        ""
      ),
      new TestCase(
        new Temp(Type.Int),
        "t1",
        "",
        ""
      ),
      new TestCase(
        new Arith(
          new Token('+'),
          new Id(new Word("x", Tag.ID), Type.Int, 4),
          new Id(new Word("y", Tag.ID), Type.Int, 4)
        ),
        "x + y",
        "",
        "\tt1 = x + y\n"
      ),
      new TestCase(
        new Unary(
          new Token('-'),
          new Id(new Word("x", Tag.ID), Type.Int, 4)
        ),
        "- x",
        "",
        "\tt1 = - x\n"
      ),
      new TestCase(
        new Access(
          new Id(new Word("arr", Tag.ID), Type.Float, 4),
          new Id(new Word("x", Tag.ID), Type.Int, 4),
          Type.Float
        ),
        "arr[ x ]",
        "",
        "\tt1 = arr[ x ]\n"
      ),
      new TestCase(
        new Or(
          Word.or,
          new Id(new Word("x", Tag.ID), Type.Bool, 4),
          new Id(new Word("y", Tag.ID), Type.Bool, 4)
        ),
        "x || y",
        "\tif x goto L3\n\tiffalse y goto L1\nL3:\tt1 = true\n\tgoto L2\nL1:\tt1 = false\nL2:",
        ""
      ),
      new TestCase(
        new And(
          Word.and,
          new Id(new Word("x", Tag.ID), Type.Bool, 4),
          new Id(new Word("y", Tag.ID), Type.Bool, 4)
        ),
        "x && y",
        "\tiffalse x goto L1\n\tiffalse y goto L1\n\tt1 = true\n\tgoto L2\nL1:\tt1 = false\nL2:",
        ""
      ),
      new TestCase(
        new Rel(
          Word.eq,
          new Id(new Word("x", Tag.ID), Type.Bool, 4),
          new Id(new Word("y", Tag.ID), Type.Bool, 4)
        ),
        "x == y",
        "\tiffalse x == y goto L1\n\tt1 = true\n\tgoto L2\nL1:\tt1 = false\nL2:",
        ""
      ),
    };

    for (var tc : Arrays.asList(testCases)) {
      Node.resetLabel();
      Temp.resetTempCount();

      assertEquals(tc.str, tc.expr.toString());

      var builder = new StringBuilder();
      tc.expr.gen(builder);
      assertEquals(tc.gen, builder.toString());

      builder.setLength(0);
      tc.expr.reduce(builder);
      assertEquals(tc.red, builder.toString());
    }
  }
}
